oak_sign/src/lib.rs

```diff
@@ -0,0 +1,169 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use anyhow::Context;
+use pem::Pem;
+use ring::{
+    rand,
+    signature::{self, Ed25519KeyPair, KeyPair},
+};
+use std::{
+    collections::HashMap,
+    fs::{read, write},
+};
+
+// PEM file tags.
+pub const PRIVATE_KEY_TAG: &str = "PRIVATE KEY";
+pub const PUBLIC_KEY_TAG: &str = "PUBLIC KEY";
+pub const SIGNATURE_TAG: &str = "SIGNATURE";
+pub const HASH_TAG: &str = "HASH";
+
+/// Convenience struct for private and public key.
+pub struct KeyPairStruct {
+    pub private_key: Vec<u8>,
+    pub public_key: Vec<u8>,
+}
+
+impl KeyPairStruct {
+    /// Generates a Ed25519 key pair.
+    pub fn generate() -> anyhow::Result<KeyPairStruct> {
+        let rng = rand::SystemRandom::new();
+        let private_key = Ed25519KeyPair::generate_pkcs8(&rng)
+            .ok()
+            .context("Couldn't generate key pair")?;
+        let key_pair = Ed25519KeyPair::from_pkcs8(private_key.as_ref())
+            .ok()
+            .context("Couldn't parse generated key pair")?;
+        let public_key = key_pair.public_key();
+        Ok(KeyPairStruct {
+            private_key: private_key.as_ref().to_vec(),
+            public_key: public_key.as_ref().to_vec(),
+        })
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct Signature {
+    pub public_key: Vec<u8>,
+    pub signature: Vec<u8>,
+    pub hash: Vec<u8>,
+}
+
+impl Signature {
+    pub fn create(input: &[u8], private_key: &[u8]) -> anyhow::Result<Signature> {
```

Comment:
Why is this accepting a byte slice for the private key? It seems more idiomatic to pass the acual parsed key pair in, which will be parsed separately by the caller, and errors handled there. The current code seems to unnecessarily couple the two.